### PR TITLE
change functional constraint functions

### PR DESCRIPTION
**Pull request by BaptisteCbl**
Hi, 
This is the pr related to the issue #113 (constraint revise).
I'll do an other pr or commits in this one for the time function revise  #115.

### CODE REVIEW

**Comment by jbcaillau** [comment]
@BaptisteCbl hi there, is this PR ready to merge? the case being I'll have a look.

**Comment by BaptisteCbl** [comment]
Hello, this PR only addresses the constraint revise, but is ready to merge in this matter per se.

File: src/model.jl [hunk 1]
```diff
@@ -801,29 +647,25 @@ Add an `:initial`, `:final`, `:control`, `:state` or `:variable` box constraint
...
+julia> constraint!(ocp, :control, rg=1, lb=0, ub=2)
 julia> constraint!(ocp, :state, rg=2:3, lb=[ 0, 0 ], ub=[ 1, 2 ])
 julia> constraint!(ocp, :initial, rg=1:2:5, lb=[ 0, 0, 0 ], ub=[ 1, 2, 1 ])
 julia> constraint!(ocp, :variable, rg=1:2, lb=[ 0, 0 ], ub=[ 1, 2 ])
 ```
 """
 function constraint!(ocp::OptimalControlModel{<: TimeDependence, <: VariableDependence}, type::Symbol; 
-    rg::Union{RangeConstraint,Nothing}=nothing, f::Union{Function,Nothing}=nothing,
-    val::Union{ctVector,Nothing}=nothing, lb::Union{ctVector,Nothing}=nothing, ub::Union{ctVector,Nothing}=nothing, 
+    rg::Union{OrdinalRange{<:Integer},Integer,Nothing}=nothing, f::Union{Function,Nothing}=nothing, lb::Union{ctVector,Nothing}=nothing, ub::Union{ctVector,Nothing}=nothing,
```
Comment (jbcaillau) [inline src/model.jl:658]: 👍🏽 thanks, that was the point

File: src/model.jl [hunk 2]
```diff
@@ -801,29 +647,25 @@ Add an `:initial`, `:final`, `:control`, `:state` or `:variable` box constraint
...
-        (::Nothing,::Nothing,::ctVector,::Nothing,::Nothing) => return constraint!(ocp, type, val, label) #
-        (::Nothing,::Nothing,::Nothing,::ctVector,::ctVector) => return constraint!(ocp, type, lb, ub, label) #
-        (::Nothing,::Function,::ctVector,::Nothing,::Nothing) => return constraint!(ocp, type, f, val, label) #
-        (::Nothing,::Function,::Nothing,::ctVector,::ctVector) => return constraint!(ocp, type, f, lb, ub, label) #
-        (::RangeConstraint,::Nothing,::ctVector,::Nothing,::Nothing) => return constraint!(ocp, type, rg, val, label) #
-        (::RangeConstraint,::Nothing,::Nothing,::ctVector,::ctVector) => return constraint!(ocp, type, rg, lb, ub, label) #
+    (typeof(rg) <: Int) && (rg = Index(rg))
+    
+    @match (rg,f,lb,ub) begin
+        (::Nothing,::Nothing,::ctVector,::ctVector) => return constraint!(ocp, type, lb, ub, label) #
```
Comment (jbcaillau) [inline src/model.jl:666]: @BaptisteCbl at this point, I see that you reduce the computation of the new `constraint!` functions with updated named arguments to three cases/subfunctions:
- linear, full range (no range give implies full range)
- linear, given range
- nonlinear (function)

Wouldn't it be more robust, particularly with less **code duplication** (*e.g.*, check on arguments), to do the converse:
- fully code the new version with named args
- implement the other calls of `constraint!` (namely the three cases mentioned above) by calling this unique new function?

File: bench/bench_nlp_constraints.md [hunk 3]
```diff
@@ -4,29 +4,30 @@
 using BenchmarkTools
```
Comment (jbcaillau) [inline bench/bench_nlp_constraints.md:4]: @BaptisteCbl just for the record: can you please recall me what these `bench` folder and `bench_nlp_constraints` are for? (did not remember this stuff.)

**Review by jbcaillau — COMMENTED** [review]
Hi @BaptisteCbl, many thanks for the work 🙏🏽 Main comment [here](https://github.com/control-toolbox/CTBase.jl/pull/116/files#r1305158620)

File: bench/bench_nlp_constraints.md (hunk 3 above)
Comment (BaptisteCbl) [reply bench/bench_nlp_constraints.md:4]: The bench folder contains programs to generate a markdown file with the performance of the Julia file. Apart from the examples that use CTBase, it is independent of CTBase. I don't think it was meant to stay here. But it was actually developed by @ocots, he may have a better answer.

**Comment by jbcaillau** [comment]
@BaptisteCbl hi again, please tell me if you have touched the request below (main comment on the `constraint!` function revision)

> Hi @BaptisteCbl, many thanks for the work 🙏🏽 Main comment [here](https://github.com/control-toolbox/CTBase.jl/pull/116/files#r1305158620). Please tell me what think.

image

**Comment by BaptisteCbl** [comment]
Hi, I haven't touched this request either but I indeed think that it would be better with your suggestion.

**Comment by BaptisteCbl** [comment]
I won't be able to do it today or this weekend, but if you want I can do it on Monday.

**Comment by jbcaillau** [comment]
> I won't be able to do it today or this weekend, but if you want I can do it on Monday.

No hurry, Monda

> I won't be able to do it today or this weekend, but if you want I can do it on Monday.

No hurry, next week is fine 👍🏽. Keep in touch.

**Comment by jbcaillau** [comment]
@BaptisteCbl having some time this Friday to work on this with you. keep me informed.

**Comment by BaptisteCbl** [comment]
Hi, I looked at the constraints! functions and tried to do as you said but I have difficulty having functions that do not overwrite themselves when we keep the same options.

**Comment by jbcaillau** [comment]
Hi @BaptisteCbl ; available for a quick catch up this Friday (Oct. 20)?

**Comment by BaptisteCbl** [comment]
Hello, yes I am available in the morning before 12h.

**Comment by jbcaillau** [comment]
> Hello, yes I am available in the morning before 12h.

👍🏽 I’ll be on zoom at 10:30 if you can
https://univ-cotedazur.zoom.us/my/caillau

**Comment by jbcaillau** [comment]
Followed by [#123]

**Pull request merged**